pennylane/hf/tapering.py

```diff
@@ -451,16 +461,155 @@ def transform_hamiltonian(h, generators, paulix_ops, paulix_sector):
+        symmstr = np.array([1 if wire in tau.wires else 0 for wire in qubit_op.wires])
+        coeff = -1 if numpy.logical_xor.reduce(numpy.logical_and(symmstr, hf_str)) else 1
+        perm.append(coeff)
+
+    return perm
+
+
+def taper_hartree_fock(num_electrons, num_wires, generators, paulix_ops, paulix_sector):
+    r"""Taper Hartree Fock according to the generated set of symmetries.
+
+    The Hartree-Fock state for a molecule is transformed to a qubit observable under Jordan-Wigner
+    transform. This observable is tapered using the same Cliffords :math:`U` that are obtained
+    from the :math:`\mathcal{Z}_2` symmetries observed in the molecular Hamiltonian. A new, tapered
+    Hartree-Fock state is built from the tapered observable by putting all the qubits which are acted
+    on by a Pauli-X or Pauli-Y operator in :math:`|1\rangle` state.
+
+    Args:
+        num_electrons (int): number of active electrons in the system for generating the Hartree-Fock bitstring
+        num_wires (int): number of wires in the system for generating the Hartree-Fock bitstring
+        generators (list[pennylane.Hamiltonian]): list of generators of symmetries, taus, for the Hamiltonian
+        paulix_ops (list[pennylane.Observable]):  list of single-qubit Pauli X operators
+        paulix_sector (list[int]): list of eigenvalues of Pauli-X operators
+
+    Returns:
+        array(int): tapered hartree-fock state :math:`|\psi\rangle_{HF}`
+
+    .. code-block:: python
+
+        >>> symbols = ['H', 'H']
+        >>> coordinates = np.array([0., 0., -0.66140414, 0., 0., 0.66140414]))
+        >>> mol = qml.hf.Molecule(symbols, coordinates)
+        >>> H, qubits = qml.hf.generate_hamiltonian(mol)(), 4
+        >>> generators, pauli_x_ops = generate_symmetries(H, qubits)
+        >>> paulix_sector, energy = find_optimal_sector(H, generators, pauli_x_ops, False, 2)
+        >>> taper_hartree_fock(2, 4, generators, pauli_x_ops, paulix_sector)
+           [1]
+    """
+
+    pauli_map = {"I": qml.Identity, "X": qml.PauliX, "Y": qml.PauliY, "Z": qml.PauliZ}
+
+    # build the untapered hartree fock state
+    hf = np.where(np.arange(num_wires) < num_electrons, 1, 0)
+
+    # convert the hf state to corresponding HF observable under JW transform
+    fermop_terms = []
+    for idx, bit in enumerate(hf):
+        if bit:
+            op_coeffs, op_str = _generate_qubit_operator([idx])
+            op_terms = []
+            for term in op_str:
+                op_term = pauli_map[term[0][1]](term[0][0])
+                for tm in term[1:]:
+                    op_term @= pauli_map[tm[1]](tm[0])
+                op_terms.append(op_term)
+            op_term = qml.Hamiltonian(np.array(op_coeffs), op_terms)
+        else:
+            op_term = qml.Hamiltonian([1], [qml.Identity(idx)])
+        fermop_terms.append(op_term)
+
+    ferm_op = fermop_terms[0]
+    for term in fermop_terms[1:]:
+        ferm_op = _observable_mult(ferm_op, term)
```

Comment:
Might use this clean code or something similar:
`ferm_op = functools.reduce(lambda i, j: qml.hf.tapering._observable_mult(i, j), fermop_terms)`  
Not sure if a new variable is needed or not. Worths thinking about it.